staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/validation/validation.go

```diff
@@ -45,6 +45,12 @@ func ValidateLabelSelectorRequirement(sr metav1.LabelSelectorRequirement, fldPat
 	case metav1.LabelSelectorOpIn, metav1.LabelSelectorOpNotIn:
 		if len(sr.Values) == 0 {
 			allErrs = append(allErrs, field.Required(fldPath.Child("values"), "must be specified when `operator` is 'In' or 'NotIn'"))
+		} else {
+			for _, v := range sr.Values {
+				for _, msg := range validation.IsValidLabelValue(v) {
+					allErrs = append(allErrs, field.Invalid(fldPath, v, msg))
+				}
```

Comment:
/hold

standard questions when we want to tighten validation:
* does it prevent a no-op update of existing persisted objects that would fail this new validation? (would block storage migration tasks)
* does it prevent deletion of existing persisted objects with finalizers that would fail this new validation? (update is required to remove finalizers to allow deletion)
* does it prevent controllers from updating status of existing persisted objects that would fail this new validation?

we could consider ratcheting validation:
* apply this tightened validation on create
* apply this tightened validation on update where the old object passed this validation (don't let a valid object become invalid)
* skip this tightened validation on update where the old object fails this validation

c.f. https://github.com/kubernetes/kubernetes/issues/59984

Reply:
Yeah +1 to these points.

Reply:
We ended up marking these pods as unschedulable back in 1.19 #93660

I wonder if we are in a better position to improve the validation in apiserver now. We could certainly ratchet the validation. 

Back in the day the question was whether it was acceptable to break the creation of ReplicaSets, but since these pods are unschedulable, I doubt that this would break any production services.